### DISCUSSION TITLE
Hopper: Distributed Fuzzer

### DISCUSSION THREAD

**Discussion by Cybergenik** [google/centipede#577] [Show and tell]
Hi,

My name is Luciano, I'm an undergrad working on a distributed fuzzer called [Hopper](https://github.com/Cybergenik/hopper). I began working on it in September 2022 after realizing the scaling limitations of AFL++ and LibFuzzer, "multi-machine" mode is just not enough. I'm also writing an undergrad Thesis on this topic, and I just recently found out that you guys open sourced this project in July 2022. I talk about Centipede in my thesis, and I'd love to get the maintainers thoughts a few questions:

- What made you want to leave the LibFuzzer project?
- Do you see Distributed Fuzzing being the next paradigm of fuzzing infrastructure?
- What makes Centipede different from other parallel/distributed attempts at fuzzing (PAFL, UltraFuzz, Hopper)?
- Why write it in C++? Why not Go or Rust?

Also, I'd love to contribute to the project! Whenever there are more issues published.

**Comment by kcc** [comment]
> * What made you want to leave the LibFuzzer project?

A number of design choices we made in libFuzzer didn't allow us to scale further. 
Mostly described in https://github.com/google/fuzztest/blob/main/centipede/doc/DESIGN.md

> * Do you see Distributed Fuzzing being the next paradigm of fuzzing infrastructure?

Paradigm isn't the right term here. 
More like requirement. 
If we want to fuzz large and slow targets, we need lots of compute and storage, and use it efficiently

> * What makes Centipede different from other parallel/distributed attempts at fuzzing (PAFL, UltraFuzz, Hopper)?

Sorry, I don't know :)

> * Why write it in C++? Why not Go or Rust?

* Language with the most mature ecosystem inside google for this kind of tasks.
* Personal preference and experience. 
* dependencies (absl is great).


> 
> Also, I'd love to contribute to the project! Whenever there are more issues published.

Stay tuned. But we have no ETA for that, sorry.
The main limitation is our ability to quickly and meaningfully review the contributions.

We just merged Centipede repo with FuzzTest. 
https://github.com/google/fuzztest
